File: skills/claude-rigor/hooks/completion_guard.js
#!/usr/bin/env node
"use strict";

// Request one skeptical review after a material code change. Read-only work and
// Workbench/scratch artifacts stay unblocked. Malformed input fails open, and
// stop_hook_active limits the review to once per turn.

const fs = require("fs");
const path = require("path");

const EDIT_TOOLS = new Set(["edit", "multiedit", "notebookedit", "write"]);
const TERMINAL_REVIEW_MARKER = "DIAGNOSIS_PROPOSAL_REVIEW_V2";
const REVIEW =
  "Before finishing, challenge the implementation once as a skeptical reviewer. " +
  "Re-read the request and inspect the final diff. Check the affected public contract, callers, " +
  "boundary and failure cases, unintended files, weakened tests, silent fallbacks, and error-path regressions. " +
  "Run the narrowest meaningful verification, inspect its actual output, fix any discovered gap, " +
  "then report only claims supported by the evidence.";

function readEntries(value) {
  if (typeof value !== "string" || value.length === 0) return [];
  try {
    return fs
      .readFileSync(value, "utf8")
      .split(/\r?\n/)
      .filter((line) => line.trim())
      .map((line) => JSON.parse(line))
      .filter((item) => item && typeof item === "object" && !Array.isArray(item));
  } catch (_) {
    return [];
  }
}

function isHumanRequest(entry) {
  if (entry.type !== "user" || entry.isMeta) return false;
  const content = entry.message && entry.message.content;
  return !(
    Array.isArray(content) &&
    content.some((item) => item && typeof item === "object" && item.type === "tool_result")
  );
}

function latestHumanRequestIndex(entries) {
  let latest = -1;
  entries.forEach((entry, index) => {
    if (isHumanRequest(entry)) latest = index;
  });
  return latest;
}

function isNonSourceArtifact(filePath, cwd) {
  if (typeof filePath !== "string" || filePath.length === 0) return false;
  const normalized = filePath.replace(/\\/g, "/").toLowerCase();
  if (
    normalized.startsWith(".workbench/") ||
    normalized.includes("/.workbench/") ||
    normalized.startsWith(".scratch/") ||
    normalized.includes("/.scratch/")
  ) return true;

  // Agents commonly place one-off probes and command output under <cwd>/tmp.
  // Those files support an investigation; they are not product-source edits.
  if (typeof cwd !== "string" || cwd.length === 0) return false;
  try {
    const resolved = path.resolve(cwd, filePath);
    const tempRoot = path.resolve(cwd, "tmp");
    return resolved === tempRoot || resolved.startsWith(tempRoot + path.sep);
  } catch (_) {
    return false;
  }
}

function hasMaterialEdit(entries, start, cwd) {
  for (const entry of entries.slice(start)) {
    if (entry.type !== "assistant") continue;
    const content = entry.message && entry.message.content;
    if (!Array.isArray(content)) continue;
    for (const item of content) {
      if (!item || typeof item !== "object" || item.type !== "tool_use") continue;
      const tool = String(item.name || "").split(".").pop().toLowerCase();
      if (!EDIT_TOOLS.has(tool)) continue;
      const input = item.input && typeof item.input === "object" ? item.input : {};
      const filePath = input.file_path || input.path || input.notebook_path;
      if (!isNonSourceArtifact(filePath, cwd)) return true;
    }
  }
  return false;
}

function isTaskWorkspace(value) {
  if (typeof value !== "string" || value.length === 0) return false;
  try {
    return fs.statSync(path.join(value, "TASK.md")).isFile();
  } catch (_) {
    return false;
  }
}

let input = "";
process.stdin.setEncoding("utf8");
process.stdin.on("data", (chunk) => {
  input += chunk;
});
process.stdin.on("end", () => {
  try {
    const payload = JSON.parse(input);
    if (!payload || typeof payload !== "object" || payload.hook_event_name !== "Stop") return;
    if (payload.stop_hook_active === true) return;
    const entries = readEntries(payload.transcript_path);
    const start = latestHumanRequestIndex(entries) + 1;
    if (entries.slice(start).some((entry) => JSON.stringify(entry).includes(TERMINAL_REVIEW_MARKER))) return;
    if (!hasMaterialEdit(entries, start, payload.cwd) && !isTaskWorkspace(payload.cwd)) return;
    process.stdout.write(JSON.stringify({ decision: "block", reason: REVIEW }));
  } catch (_) {
    // A guard that cannot understand its own input must not prevent completion.
  }
});

File: claude_harness_eval/hooks/completion_guard.py
#!/usr/bin/env python3
"""Request one skeptical completion review after Claude edits code.

This is a Claude Code Stop hook. It fails open on malformed input and honors
``stop_hook_active`` so the review can happen at most once per turn.
"""

from __future__ import annotations

import json
from pathlib import Path
import sys
from typing import Any


EDIT_TOOLS = {"edit", "multiedit", "notebookedit", "write"}
REVIEW = (
    "Before finishing, challenge the implementation once as a skeptical reviewer. "
    "Re-read the request and inspect the final changes. Build explicit input partitions "
    "for valid, boundary, malformed, wrong-type, and failure cases; compare every partition "
    "with the required public contract and existing behavior. Resolve any stated judgment calls "
    "from repository evidence instead of narrowing the contract or asking the user. Check for "
    "missed callers, unintended files, weakened tests, silent fallbacks, and error-path regressions. "
    "Run the narrowest meaningful verification available, fix any discovered gap, then report only "
    "claims supported by executed evidence."
)


def _read_entries(value: Any) -> list[dict[str, Any]]:
    if not isinstance(value, str) or not value:
        return []
    path = Path(value)
    if not path.is_file():
        return []
    entries: list[dict[str, Any]] = []
    try:
        for line in path.read_text(encoding="utf-8", errors="replace").splitlines():
            if not line.strip():
                continue
            item = json.loads(line)
            if isinstance(item, dict):
                entries.append(item)
    except (OSError, json.JSONDecodeError):
        return []
    return entries


def _is_human_request(entry: dict[str, Any]) -> bool:
    if entry.get("type") != "user" or entry.get("isMeta"):
        return False
    content = (entry.get("message") or {}).get("content")
    if isinstance(content, list) and any(
        isinstance(item, dict) and item.get("type") == "tool_result"
        for item in content
    ):
        return False
    return True


def _edited_since_latest_request(entries: list[dict[str, Any]]) -> bool:
    start = 0
    for index, entry in enumerate(entries):
        if _is_human_request(entry):
            start = index + 1
    for entry in entries[start:]:
        if entry.get("type") != "assistant":
            continue
        content = (entry.get("message") or {}).get("content")
        if not isinstance(content, list):
            continue
        for item in content:
            if not isinstance(item, dict) or item.get("type") != "tool_use":
                continue
            tool = str(item.get("name") or "").rsplit(".", 1)[-1].lower()
            if tool in EDIT_TOOLS:
                return True
    return False


def _is_task_workspace(value: Any) -> bool:
    if not isinstance(value, str) or not value:
        return False
    try:
        return (Path(value) / "TASK.md").is_file()
    except OSError:
        return False


def main() -> int:
    try:
        payload = json.load(sys.stdin)
    except (json.JSONDecodeError, OSError):
        return 0
    if not isinstance(payload, dict) or payload.get("hook_event_name") != "Stop":
        return 0
    if payload.get("stop_hook_active") is True:
        return 0
    edited = _edited_since_latest_request(_read_entries(payload.get("transcript_path")))
    if not edited and not _is_task_workspace(payload.get("cwd")):
        return 0
    print(json.dumps({"decision": "block", "reason": REVIEW}, separators=(",", ":")))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
